main.nf

```diff
@@ -167,46 +486,759 @@ process fastqc {
+    output:
+    file("taxonomy/taxonomy.tab") into file_krona_db
+
+    when:
+    ( params.centrifuge_db || params.kraken2_db ) && !params.skip_krona
+
+    script:
+    """
+    ktUpdateTaxonomy.sh taxonomy
+    """
+}
+
+centrifuge_to_krona
+    .mix(kraken2_to_krona)
+    .combine(file_krona_db)
+    .set { krona_input }
+
+process krona {
+    tag "${classifier}-${name}"
+    publishDir "${params.outdir}/Taxonomy/${classifier}/${name}", mode: 'copy'
+
+    input:
+    set val(classifier), val(name), file(report), file("taxonomy/taxonomy.tab") from krona_input
+
+    output:
+    file("*.html")
+
+    script:
+    """
+    ktImportTaxonomy "$report" -tax taxonomy
+    """
+}
+
+
+/*
+ * STEP 2 - Assembly
+ */
+process megahit {
+    tag "$name"
+    publishDir "${params.outdir}/", mode: 'copy',
+        saveAs: {filename -> filename.indexOf(".fastq.gz") == -1 ? "Assembly/$filename" : null}
+
+    input:
+    set val(name), file(reads) from trimmed_reads_megahit
+
+    output:
+    set val("MEGAHIT"), val("$name"), file("MEGAHIT/${name}.contigs.fa") into assembly_megahit_to_quast
+    set val("MEGAHIT"), val("$name"), file("MEGAHIT/${name}.contigs.fa") into assembly_megahit_to_metabat
+    file("MEGAHIT/*.log")
+
+    when:
+    !params.skip_megahit
+
+    script:
+    if ( !params.singleEnd ) {
+    """
+    megahit -t "${task.cpus}" -1 "${reads[0]}" -2 "${reads[1]}" -o MEGAHIT \
+        --out-prefix "${name}"
+    """
+    }
+    else {
+    """
+    megahit -t "${task.cpus}" -r ${reads} -o MEGAHIT --out-prefix "${name}"
+    """
+    }
+}
+
+
+/*
+ * metaSpades hybrid Assembly
+ */
+
+ files_lr_filtered
+    .combine(trimmed_sr_spadeshybrid, by: 0)
+    .set { files_pre_spadeshybrid }
+
+process spadeshybrid {
+    tag "$id"
+    publishDir "${params.outdir}/", mode: 'copy', pattern: "${id}*",
+        saveAs: {filename -> filename.indexOf(".fastq.gz") == -1 ? "Assembly/SPAdesHybrid/$filename" : null}
+
+    input:
+    set id, file(lr), file(sr) from files_pre_spadeshybrid  
+
+    output:
+    set id, val("SPAdesHybrid"), file("${id}_graph.gfa") into assembly_graph_spadeshybrid
+    set val("SPAdesHybrid"), val("$id"), file("${id}_scaffolds.fasta") into assembly_spadeshybrid_to_quast
+    set val("SPAdesHybrid"), val("$id"), file("${id}_scaffolds.fasta") into assembly_spadeshybrid_to_metabat
+    file("${id}_contigs.fasta")
+    file("${id}_log.txt")
+
+    when:
+    params.manifest && !params.singleEnd && !params.skip_spadeshybrid
+     
+    script:
+    def maxmem = "${task.memory.toString().replaceAll(/[\sGB]/,'')}"
+    """
+    metaspades.py \
+        --threads "${task.cpus}" \
+        --memory "$maxmem" \
+        --pe1-1 ${sr[0]} \
+        --pe1-2 ${sr[1]} \
+        --nanopore ${lr} \
+        -o spades
+    cp spades/assembly_graph_with_scaffolds.gfa ${id}_graph.gfa
+    cp spades/scaffolds.fasta ${id}_scaffolds.fasta
+    cp spades/contigs.fasta ${id}_contigs.fasta
+    cp spades/spades.log ${id}_log.txt
```

Comment:
Why do you need to copy these? Can you not just specify their original paths in the output?

Reply:
There are no options in SPAdes to change the prefix of the output files. If we have several samples there will be conflicting names. or you mean I should rename on the fly in the publishDir code? (using a saveAs closure to add a prefix to all the files felt complicated)

Reply:
Renaming in the `script` chunk felt easier because then we can catch the four files we want with `pattern: "${id}*"` in the `publishDir`, not sure how to combine that with renaming with `saveAs`

Reply:
I’ll try to write a suggestion (or maybe someone else can). `saveAs` for publishDir and can use a var alongside the files in the output channels to organise into folders in downstream processes if needed.

Reply:
I could `mv` instead of `copy` though, as seen in the assembly step of the `bacass` pipeline

Reply:
Are you sure that the `publishDir` thing is working? To me, it looks like it is setting the filename for anything that doesn't contain `".fastq.gz"` to `null`, which means that it won't be saved.. My expectation is that _only_ the `fastq.gz` files are being saved to the results directory here.

Reply:
Added example code snippets above. Hopefully that all makes sense, let me know if it doesn't!

Alternatively, feel free to ignore me and tell me to make my own PR instead of hassling you..